scene/gui/layout_container.cpp

```diff
@@ -0,0 +1,236 @@
+	ADD_PROPERTYI(PropertyInfo(Variant::INT, "margin_right", PROPERTY_HINT_RANGE, "-4096,4096,suffix:px"), "set_margin", "get_margin", SIDE_RIGHT);
+	ADD_PROPERTYI(PropertyInfo(Variant::INT, "margin_bottom", PROPERTY_HINT_RANGE, "-4096,4096,suffix:px"), "set_margin", "get_margin", SIDE_BOTTOM);
+
+	BIND_ENUM_CONSTANT(PRESET_TOP_LEFT);
+	BIND_ENUM_CONSTANT(PRESET_TOP_RIGHT);
+	BIND_ENUM_CONSTANT(PRESET_BOTTOM_LEFT);
+	BIND_ENUM_CONSTANT(PRESET_BOTTOM_RIGHT);
+	BIND_ENUM_CONSTANT(PRESET_CENTER_LEFT);
+	BIND_ENUM_CONSTANT(PRESET_CENTER_TOP);
+	BIND_ENUM_CONSTANT(PRESET_CENTER_RIGHT);
+	BIND_ENUM_CONSTANT(PRESET_CENTER_BOTTOM);
+	BIND_ENUM_CONSTANT(PRESET_CENTER);
+	BIND_ENUM_CONSTANT(PRESET_LEFT_WIDE);
+	BIND_ENUM_CONSTANT(PRESET_TOP_WIDE);
+	BIND_ENUM_CONSTANT(PRESET_RIGHT_WIDE);
+	BIND_ENUM_CONSTANT(PRESET_BOTTOM_WIDE);
+	BIND_ENUM_CONSTANT(PRESET_VCENTER_WIDE);
+	BIND_ENUM_CONSTANT(PRESET_HCENTER_WIDE);
+	BIND_ENUM_CONSTANT(PRESET_FULL_RECT);
+	BIND_ENUM_CONSTANT(PRESET_CUSTOM);
+
+	BIND_THEME_ITEM(Theme::DATA_TYPE_CONSTANT, LayoutContainer, margin_left);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_CONSTANT, LayoutContainer, margin_top);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_CONSTANT, LayoutContainer, margin_right);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_CONSTANT, LayoutContainer, margin_bottom);
+}
+
+Size2 LayoutContainer::get_minimum_size() const {
+	Size2 ms;
+	for (int i = 0; i < get_child_count(); i++) {
+		Control *c = as_sortable_control(get_child(i));
+		if (!c) {
+			continue;
+		}
+		Size2 minsize = c->get_combined_minimum_size();
+		ms = ms.max(minsize);
+	}
+	ms += Vector2(margins[SIDE_LEFT] + margins[SIDE_RIGHT], margins[SIDE_TOP] + margins[SIDE_BOTTOM]);
+	ms += Vector2(theme_cache.margin_left + theme_cache.margin_right, theme_cache.margin_top + theme_cache.margin_bottom);
+	ms = ms.max(Vector2());
+	return ms;
+}
+
+void LayoutContainer::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_SORT_CHILDREN: {
+			_resort();
+		} break;
+
+		case NOTIFICATION_THEME_CHANGED: {
+			update_minimum_size();
+		} break;
+
+		case NOTIFICATION_TRANSLATION_CHANGED:
+		case NOTIFICATION_LAYOUT_DIRECTION_CHANGED: {
+			queue_sort();
+		} break;
+	}
+}
+
+void LayoutContainer::_validate_property(PropertyInfo &p_property) const {
+	if (p_property.name.begins_with("custom_relative_margin_") && preset != LayoutContainerPreset::PRESET_CUSTOM) {
+		p_property.usage = PROPERTY_USAGE_NO_EDITOR;
+	}
+}
+
+Vector<int> LayoutContainer::get_allowed_size_flags_horizontal() const {
+	Vector<int> flags;
+	return flags;
+}
+
+Vector<int> LayoutContainer::get_allowed_size_flags_vertical() const {
+	Vector<int> flags;
+	return flags;
+}
+
+void LayoutContainer::set_margin_layout_preset(LayoutContainer::LayoutContainerPreset p_preset) {
+	if (p_preset == preset) {
+		return;
+	}
+
+	preset = p_preset;
+	notify_property_list_changed();
```

Comment:
It would be more efficient and user-friendly to call this only when changing between custom and non-custom.